Fix pagination.

markrcote opened the pull request:
`Client#paginate` expects the returned `data` to be an Array, but non-paginated API calls (e.g. `build`) return only a Resource object.

Another way to fix this would be to have paginated (or non-paginated) calls explicitly identify themselves as such, rather than having the `paginate` call determine this based on the class of the returned `data`, but this might be annoying to maintain.

casperisfine commented:
> Another way to fix this would be to  [..] non-paginated calls explicitly identify themselves as such,

Seems better on paper.

casperisfine commented:
ALso do you have an example of the issue? It's been a while I worked on `buildkit` but I don't quite see when this happens.

markrcote commented:
This seems to be pretty unmaintained, but I'd like to use the pagination functionality, which appears to be broken.

casperisfine commented:
> This seems to be pretty unmaintained

It's maintained. It simply didn't need changes nor got reported bug fixes for a while.

> which is partly broken.

I still don't understand what the problem is. Mind sharing a snippet or even a test case ?

markrcote commented:
> > This seems to be pretty unmaintained
> 
> It's maintained. It simply didn't need changes nor got reported bug fixes for a while.

Sorry, no offence intended. :) I just meant that it still uses Travis CI, and those tests are now broken.

> > which is partly broken.
> 
> I still don't understand what the problem is. Mind sharing a snippet or even a test case ?

It's pretty easy to replicate. Any call to `build()` fails if `auto_paginate` is enabled, e.g.

```
require "buildkit"

client = Buildkit.new(
  token: <token>,
  auto_paginate: true
)

build = client.build(
  "shopify",
  "shipit-mobile-testing-android-build",
  226
)
```

```
$ bundler exec ruby test.rb
Traceback (most recent call last):
	6: from test.rb:8:in `<main>'
	5: from /Users/mcote/src/github.com/Shopify/buildkit/lib/buildkit/client/builds.rb:50:in `build'
	4: from /Users/mcote/src/github.com/Shopify/buildkit/lib/buildkit/client.rb:57:in `get'
	3: from /Users/mcote/src/github.com/Shopify/buildkit/lib/buildkit/client.rb:140:in `paginate'
	2: from /Users/mcote/src/github.com/Shopify/buildkit/lib/buildkit/client.rb:140:in `loop'
	1: from /Users/mcote/src/github.com/Shopify/buildkit/lib/buildkit/client.rb:142:in `block in paginate'
/Users/mcote/src/github.com/Shopify/buildkit/lib/buildkit/client.rb:142:in `concat': can't convert Sawyer::Resource to Array (Sawyer::Resource#to_ary gives NilClass) (TypeError)
```

A call to an API which returns multiple items, e.g. `artifacts()`, works fine.

markrcote commented:
> > > This seems to be pretty unmaintained
> > 
> > 
> > It's maintained. It simply didn't need changes nor got reported bug fixes for a while.
> 
> Sorry, no offence intended. :) I just meant that it still uses Travis CI, and those tests are now broken.

And I just realized we use Travis because this is open source, so forget that part. :)

markrcote commented:
> Ok make sense. We indeed never really used the autopaginate feature, it's useful for small scripts, but really discouraged for production use.

Why is this? Because the way the response is mangled might be unexpected/confusing?

> I cleaned up master, if you rebase you should have a green CI.
> 
> Also a unit test would be nice.

Added a test that verify that `build` calls still work when auto_pagination is enabled.

casperisfine commented:
> > it's useful for small scripts, but really discouraged for production use.
> 
> Why is this?

Mostly because if you happen to follow a big list by accident (say's `shopify/shopify` builds list) your production code my hang for a very long time and loads tons of data in memory.

It's usually preferred to explicitly paginate where you expect it.

Also that option is global which can surprise people. It's there because `buildkit` mimick Octokit and Octokit have it, but really I regret including it.

Doesn't mean it shouldn't be fixed though. If it's there it should work.

casperisfine commented on spec/client/builds_spec.rb:
```diff
@@ -50,6 +50,17 @@
+  context '#paginated_build' do
+    it 'returns the build' do
+      VCR.use_cassette 'build' do
+        client.auto_paginate = true
```
Just make sure that client is instantiated new for every spec. Otherwise you'll leak that state.

markrcote replied:
Yup I verified that it is.
